### PR TITLE
[common-library] Add common library and a template chart

### LINKED ISSUE #722: [common-library] Finish the common library to use in the other charts
To give consistency through all the charts we need to create a library chart that has been tested to work.

- [ ] Create the common library
- [ ] Create a chart template to be able to copy it and test the common library

### PR DESCRIPTION
**Pull request by kang-makes**
#### Is this a new chart

Yes

#### What this PR does / why we need it:

It separates the common logic from other charts to be used as a consistency measure between all the charts.

#### Which issue this PR fixes
 - Fixes #722 

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ ] Chart Version bumped
- [ ] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[mychartname]`)

### CODE REVIEW

**Comment by kang-makes** [comment]
```
Installing chart 'common-library => (version: "0.1.0", path: "charts/common-library")'...
Error: INSTALLATION FAILED: library charts are not installable
```

We will have to change the tests or create another folder for the library charts so we don't try to install them.

We would have also to find a way to test this chart.

File: .github/workflows/lint_test_charts.yaml [hunk 1]
```diff
@@ -31,34 +31,42 @@ jobs:
...
         uses: helm/chart-testing-action@v2.1.0
 
+      - name: Set up helm-unittest
+        run: |
+          helm plugin install https://github.com/quintush/helm-unittest
+
       - name: Lint charts
-        run: ct lint --config ct.yaml
+        run: |
+          for realm in library charts; do
```
[the remark this answers is not in the document]
Comment (paologallinaharbur) [reply .github/workflows/lint_test_charts.yaml:40]: Isn't easier just to call the "ct lint" twice?

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -45,10 +44,20 @@ jobs:
           git config user.name "$GITHUB_ACTOR"
           git config user.email "$GITHUB_ACTOR@users.noreply.github.com"
 
-      - name: Install Helm
-        uses: azure/setup-helm@v1
+      # As of today, there is no way to filter/exclude charts to be released. Config does not support it:
+      # https://github.com/helm/chart-releaser/blob/4eb598f96aa53d1c0b8234af2087834e1f2275e1/pkg/config/config.go#L40-L61
+      - name: remove the chart template as it should not be deployed
+        run: |
+          rm -rf library/CHART-TEMPLATE
```
[the remark this answers is not in the document]
Comment (paologallinaharbur) [reply .github/workflows/release.yml:51]: Maybe deleting only the chart.yaml? https://github.com/helm/chart-releaser-action/pull/18

File: library/CHART-TEMPLATE/Chart.yaml [hunk 3]
```diff
@@ -0,0 +1,47 @@
+apiVersion: v2
+name: CHART-TEMPLATE
```
Comment (paologallinaharbur) [inline library/CHART-TEMPLATE/Chart.yaml:2]: the name is uppercase to avoid it to be installable?

File: library/CHART-TEMPLATE/Chart.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+description: A Helm chart for Kubernetes
```
Comment (paologallinaharbur) [inline library/CHART-TEMPLATE/Chart.yaml:3]: I would maybe add few lines saying that this chart is just an example to create unit tests for the common library

File: library/common-library/Chart.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+apiVersion: v2
+name: common-library
+description: Provides helpers to provide consistency on all the charts
```
Comment (paologallinaharbur) [inline library/common-library/Chart.yaml:3]: This sentence could be improved a bit

File: library/common-library/templates/_staging.tpl [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+{{- else -}}
+{{- /* This allows us to use `$global` as an empty dict directly in case `Values.global` does not exists */ -}}
+{{- $global := index .Values "global" | default dict -}}
+{{- if get $global "nrStaging" | kindIs "bool" -}}
+    {{- if $global.nrStaging -}}
+        {{- $global.nrStaging -}}
+    {{- end -}}
+{{- end -}}
+{{- end -}}
+{{- end -}}
```
Comment (paologallinaharbur) [inline library/common-library/templates/_staging.tpl:26]: There is a "library/common-library/values.yaml" committed, do we need it for the CI, linting, etc?
If so I would keep it, but I would add a comment saying why it is empty and why we need it

File: library/common-library/templates/_license-test-pod.yaml [hunk 7]
```diff
@@ -0,0 +1,25 @@
+{{- define "common.tests.license" -}}
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  annotations:
+    helm.sh/hook: test-success
+    helm.sh/hook-delete-policy: hook-succeeded
+  labels:
```
Comment (paologallinaharbur) [inline library/common-library/templates/_license-test-pod.yaml:9]: didn't we say we were getting rid of this since it was not straightforward to do so?

File: library/CHART-TEMPLATE/Chart.yaml (hunk 3 above)
Comment (kang-makes) [reply library/CHART-TEMPLATE/Chart.yaml:2]: It is uppercase to be easier to `grep CHART-TEMPLATE` or `sed -i s_CHART-TEMPLATE_whatever_` and to catch attention while templating it.

File: library/CHART-TEMPLATE/values.yaml [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  #   - ALL
+  # readOnlyRootFilesystem: true
+  # runAsNonRoot: true
+  # runAsUser: 1000
+
+service:
+  type: ClusterIP
+  port: 80
+
+### Not implemented yet
```
[the remark this answers is not in the document]
Comment (kang-makes) [reply library/CHART-TEMPLATE/values.yaml:54]: Those were notes I left for me during the POC that I didn't revisit. Sorry.

File: library/common-library/templates/_low-data-mode.tpl [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+Abstraction of the lowDataMode toggle.
+This helper allows to override the global `.global.lowDataMode` with the value of `.lowDataMode`.
+Returns "true" if `lowDataMode` is enabled, otherwise "" (empty string)
+*/ -}}
+{{- define "common.lowDataMode" -}}
+{{- /* `get` will return "" (empty string) if value is not found, and the value otherwise, so we can type-assert with kindIs */ -}}
+{{- if (get .Values "lowDataMode" | kindIs "bool") -}}
+    {{- if .Values.lowDataMode -}}
+        {{- /*
+            We want only to return when this is true, returning `false` here will template "false" (string) when doing
```
[the remark this answers is not in the document]
Comment (kang-makes) [reply library/common-library/templates/_low-data-mode.tpl:11]: It is the current implementation that is in the charts: https://github.com/newrelic/helm-charts/blob/master/charts/nri-prometheus/templates/_helpers.tpl#L118-L141

File: .gitignore [hunk 10]
```diff
@@ -35,7 +38,3 @@ Session.vim
 .netrwhist
 
 .history
-
-/values.yaml
-
-__snapshot__
```
[the remark this answers is not in the document]
Comment (paologallinaharbur) [reply .gitignore:41]: yes, but do we what that in git? or are we excluding it somehow?

File: .github/ct-charts.yaml [hunk 11]
```diff
@@ -1,4 +1,3 @@
-# configuration file for https://github.com/helm/chart-testing-action
```
[the remark this answers is not in the document]
Comment (kang-makes) [reply .github/ct-charts.yaml]: Solved in #cd009b8b9c519f4ce67907e67267414b4e5e09de

File: library/common-library/templates/_staging.tpl (hunk 6 above)
Comment (paologallinaharbur) [reply library/common-library/templates/_staging.tpl:26]: can we remote this or add a comment? just to avoid having a completely empty file "flying" 
Screen Shot 2022-03-02 at 11 03 39 AM

File: .github/workflows/release.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (kang-makes) [reply .github/workflows/release.yml]: Solved in [#cd009b8](https://github.com/newrelic/helm-charts/pull/713/commits/cd009b8b9c519f4ce67907e67267414b4e5e09de)

File: library/common-library/templates/_license-test-pod.yaml (hunk 7 above)
Comment (kang-makes) [reply library/common-library/templates/_license-test-pod.yaml]: Solved in [#cd009b8](https://github.com/newrelic/helm-charts/pull/713/commits/cd009b8b9c519f4ce67907e67267414b4e5e09de)

File: library/CHART-TEMPLATE/test/common-library-test--serviceaccount-name.yaml [hunk 12]
```diff
@@ -0,0 +1,46 @@
+suite: test service account naming
```
Comment (paologallinaharbur) [inline library/CHART-TEMPLATE/test/common-library-test--serviceaccount-name.yaml:1]: tests needs to be suffixed by _test to be executed automatically without -f

File: library/CHART-TEMPLATE/Chart.yaml (hunk 3 above)
Comment (paologallinaharbur) [reply library/CHART-TEMPLATE/Chart.yaml:2]: fine for me!

File: .gitignore (hunk 10 above)
[the remark this answers is not in the document]
Comment (paologallinaharbur) [reply .gitignore:41]: makes sense

File: library/common-library/templates/_staging.tpl (hunk 6 above)
Comment (kang-makes) [reply library/common-library/templates/_staging.tpl:26]: It is needed for `helm lint` to work.

File: library/CHART-TEMPLATE/test/common-library-test--serviceaccount-name.yaml (hunk 12 above)
Comment (kang-makes) [reply library/CHART-TEMPLATE/test/common-library-test--serviceaccount-name.yaml]: Done in [#cc90065](https://github.com/newrelic/helm-charts/pull/713/commits/cc9006581557b1cfdc6a73e7f35e94fd69a92bf4)

**Comment by paologallinaharbur** [comment]
LGTM the only doubt I have it is regarding having charts in separate folders. Bitnami for example is keeping them in the charts folder has well https://github.com/bitnami/charts/tree/master/bitnami/common

File: .github/ct-lint.yaml [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+# ```yaml
+# chart-dirs:
+#   - library
+#   - charts
+# ```
+
+# In that case `ct install` is going to try to install the library chart and is going to fail. That is the reason there are two `ct.yaml` files, one
+# for the step that lint that includes all the folders at the same time and other of the installation step that test installability
+
+chart-dirs:
```
Comment (paologallinaharbur) [inline .github/ct-lint.yaml:13]: Could we have them in the same directory and add with the exclude https://github.com/bitnami/charts/blob/596d26b70b69b438e15047e83710345ee49f8f57/.github/ct-install.yaml#L6

to the ct-install?

Comment (kang-makes) [reply .github/ct-lint.yaml:13]: I said that we could not do that because I did not read the source code properly: https://github.com/helm/chart-testing/blob/main/pkg/config/config.go#L59

On the other hand, I would not change it because if we do it is going to try to install the `CHART-TEMPLATE` which is going to fail. I would also want to lint `common-library` so pipeline tests if we have properly bumped the version and such.

Good catch nevertheless...

**Comment by kang-makes** [comment]
There is no idiomatic way of having the charts stored.

Bitnami has everything in the same folder but another example is k8s-at-home which has [a different repository for libraries](https://github.com/k8s-at-home/library-charts) and they separate workloads for [stable](https://github.com/k8s-at-home/charts/tree/master/charts/stable) and [incubator](https://github.com/k8s-at-home/charts/tree/master/charts/incubator). 

I would prefer to keep them separated into two different folders because we are treating them differently. One folder is for workloads that are being deployed while another folder is simply tooling.

Let's wait until @roobre or @gsanchezgavier say something about this if you wish.

File: .github/ct-lint.yaml (hunk 13 above)
Comment (kang-makes) [reply .github/ct-lint.yaml:13]: Oh, and the one that has no exceptions is the chart release: https://github.com/helm/chart-releaser/blob/main/pkg/config/config.go#L40-L61

So it is not that I did not read properly but that in my mind we are talking about the same tool :)

Comment (paologallinaharbur) [reply .github/ct-lint.yaml:13]: I am fine with dividing them if it is not a standard to keep them under charts
If we want to keep the same we could do the following:

One for the install
```
exclude
 - common-library
 - CHART-TEMPLATE
chart-repos:
  - kube-state-metrics=https://kubernetes.github.io/kube-state-metrics
  - pixie-operator=https://pixie-operator-charts.storage.googleapis.com
```

One for the linting
```
chart-repos:
  - kube-state-metrics=https://kubernetes.github.io/kube-state-metrics
  - pixie-operator=https://pixie-operator-charts.storage.googleapis.com
```

File: .github/workflows/release.yml [hunk 14]
```diff
@@ -45,10 +44,21 @@ jobs:
...
-        uses: azure/setup-helm@v1
+      # As of today, there is no way to filter/exclude charts to be released. Config does not support it:
+      # https://github.com/helm/chart-releaser/blob/4eb598f96aa53d1c0b8234af2087834e1f2275e1/pkg/config/config.go#L40-L61
+      # So we delete the Chart.yaml to filter the chart: https://github.com/helm/chart-releaser-action/pull/18
+      - name: remove the chart template as it should not be deployed
+        run: |
+          rm ./library/CHART-TEMPLATE/Chart.yaml
+
+      - name: Release library charts
+        uses: helm/chart-releaser-action@v1.2.1
```
Comment (gsanchezgavier) [inline .github/workflows/release.yml:55]: https://github.com/helm/chart-releaser-action/releases/tag/v1.3.0
maybe is worth to bump it.

File: library/common-library/templates/_low-data-mode.tpl [hunk 15]
```diff
@@ -0,0 +1,26 @@
+{{- /*
+Abstraction of the lowDataMode toggle.
+This helper allows to override the global `.global.lowDataMode` with the value of `.lowDataMode`.
+Returns "true" if `lowDataMode` is enabled, otherwise "" (empty string)
+*/ -}}
+{{- define "common.lowDataMode" -}}
+{{- /* `get` will return "" (empty string) if value is not found, and the value otherwise, so we can type-assert with kindIs */ -}}
+{{- if (get .Values "lowDataMode" | kindIs "bool") -}}
```
[the remark this answers is not in the document]
Comment (kang-makes) [reply library/common-library/templates/_low-data-mode.tpl:8]: This logic is also used for [creating the service account](https://github.com/newrelic/helm-charts/pull/713/files#diff-4b3d46fb917e8d56b8de51a27e90a24a9a0eff7215b74052d67360f8483730e2R1-R28) and [it is tested](https://github.com/newrelic/helm-charts/pull/713/files#diff-37a891818928aebca3dd07fc22b73a546722dc6c509067e4f9cfb7cfb9f6a028)

But you are completely right, that test should exist too. What I don't know is how because it should template something like an example of configMap. Do you have any idea of an example you would like to have?

File: library/CHART-TEMPLATE/tests/common_library_naming_test.yaml [hunk 16]
```diff
@@ -0,0 +1,56 @@
+suite: test naming helpers
+templates:
+  - templates/deployment.yaml
+  - templates/service.yaml
+release:
+  name: my-release
+  namespace: my-namespace
+set:
+  licenseKey: test
+  cluster: test
```
[the remark this answers is not in the document]
Comment (kang-makes) [reply library/CHART-TEMPLATE/tests/common_library_naming_test.yaml]: Solved in [d11b6ad](https://github.com/newrelic/helm-charts/pull/713/commits/d11b6ad0e194de9fe72561553c8986158bf11cf9)

File: .github/workflows/release.yml (hunk 14 above)
Comment (kang-makes) [reply .github/workflows/release.yml]: Solved in [d11b6ad](https://github.com/newrelic/helm-charts/pull/713/commits/d11b6ad0e194de9fe72561553c8986158bf11cf9)

File: .github/ct-install.yaml [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+# ```
+
+# In that case `ct install` is going to try to install the library chart and is going to fail. That is the reason there are two `ct.yaml` files, one
+# for the step that lint that includes all the folders at the same time and other of the installation step that test installability
+
+chart-dirs:
+# - library  # left here this comment on purpose for clarity' sake
+  - charts
+
+chart-repos:
```
[the remark this answers is not in the document]
Comment (kang-makes) [reply .github/ct-install.yaml]: Solved in [d11b6ad](https://github.com/newrelic/helm-charts/pull/713/commits/d11b6ad0e194de9fe72561553c8986158bf11cf9)

File: .github/workflows/release.yml [hunk 18]
```diff
@@ -30,9 +30,8 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v2
-
-      - name: Fetch history
-        run: git fetch --prune --unshallow
+        with:
+          fetch-depth: 0
```
[the remark this answers is not in the document]
Comment (kang-makes) [reply .github/workflows/release.yml:35]: Solved in [d11b6ad](https://github.com/newrelic/helm-charts/pull/713/commits/d11b6ad0e194de9fe72561553c8986158bf11cf9)

File: library/common-library/templates/_low-data-mode.tpl (hunk 15 above)
[the remark this answers is not in the document]
Comment (gsanchezgavier) [reply library/common-library/templates/_low-data-mode.tpl:8]: I would add the most independent and short template to test this logic , like an unit test. But what i mean was more general , let me rephrase it , are we going to have unit test on the library chart?

**Review by paologallinaharbur — APPROVED** [review]
